### PR TITLE
[docs] Added documentation for memory management

### LINKED ISSUE #945: Memory optimization wiki
There have been a lot of memory optimizations added to nntrainer, and different semantics to handle memory management.
Summarize all the information appropriately in corresponding headers or in wiki.

### PR DESCRIPTION
**Pull request by kparichay**
Added documentation for memory management.

Resolves #945

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: docs/memory-management.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+- **Weights** which make the model
+- **Input, Output and Label tensors** for the model
+- **Variables** required to run the model like intermediate inputs and outputs, derivatives, etc.
+
+The management of the memory for the above 3 kinds of memory is done depending on the mode of execution of the model, namely `inference` or `training` mode. The memory management of these tensors is performed by the `Manager` inside `NNTrainer`.
+
+The memory `weights` are allocated at the time of initializing the model, where the `weights` are either initialized using the set initializer or loaded from the saved model file. The memory allocated for `weights` is freed
+upon `destruction` of the model object containing the `weights`.
+
+The memory for the `Input, Output and Label tensors` and `Variables` is allocated lazily, and de-allocated once its usage is finished.
```
Comment (zhoonit) [inline docs/memory-management.md:18]: May be this?
```suggestion
The memory for the `Input, Output and Label tensors` and `Gradients` is allocated lazily, and de-allocated once its usage is finished.
```

Comment (kparichay) [reply docs/memory-management.md:18]: It should be `variables` IMO.

Comment (zhoonit) [reply docs/memory-management.md]: Right. I checked the definition

File: docs/memory-management.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+The memory allocation for weights is done independently of the mode of execution for the model. The total memory allocated for weights will be equal or smaller than the saved model binary size (as saved model binary size can include training-related parameters).
+
+## Input, Output, and Label Tensors
+
+The input, output, and label tensors of the model are managed differently based on the mode of execution of the model.
+
+- In `training` mode, `Manager` allocates the memory of the input, output, and the label tensors preemptively. These pre-allocated tensors are connected with the dataset and reused in each iteration for memory-efficient training.
+- In `inference` mode, as both input and label are allocated and provided by the user, `Manager` preemptively allocates the memory only for output tensors of the model. The input and label tensors provided by the users are directly used for inference.
```
Comment (zhoonit) [inline docs/memory-management.md:36]: How about adding the statement that user should guarantee that the input and label must remain valid and unchanged until the inference is finished.

File: docs/memory-management.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+Gradients are allocated only while `training` the model. The memory required by the gradients is optimized by the `Manager` by using a shared tensor. A shared tensor with the size of the maximum memory required by a weight is allocated, and all the gradients use this shared tensor for their memory requirements.
+
+This optimization comes with the cost that each weight must be updated by its gradient before calculating the gradient for the next weight. Further, the gradient of an individual weight is not available after an iteration.
+
+If multiple weights in a layer require all their gradients to be calculated before being applied, the optimization can increase the size of the shared tensor to the size of maximum memory required by all the weights of a layer in the model (This is not yet supported).
+
+### Intermediate Inputs and Outputs
+
+The allocations for input and output tensors between layers are optimized dependeninh on the execution mode of the model.
```
Comment (zhoonit) [inline docs/memory-management.md:52]:
```suggestion
The allocations for input and output tensors between layers are optimized dependent on the execution mode of the model.
```

File: docs/memory-management.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+# Memory Optimizations using In-Place Layer Operations
+
+Certain layers can operate in-place, allowing more memory optimizations to be performed. Layers performing in-place operations do not require memory for storing their input(s), and their output(s) overwrite their inputs. The back-propagation operation for such layers also must not require their inputs.
+
+- Batch Normalization layer can operate in-place. BN layer stores processed input rather than the original input to achieve higher performance. This allows the back-propagation operation to work without inputs, and for BN layer to work in-place.
+
+- Various activation Layers can also operate in-place (limited to `ReLU`, `Sigmoid`, `TanH`, etc. operations that work on individual elements and do not depend on their neighboring values). These activation layers do not need to store their inputs and rather depend on their outputs for faster performance. Such layers are optimized to work in-place. Note that derivatives for activation layers do not overwrite their outputs as explained in the previous section.
```
Comment (zhoonit) [inline docs/memory-management.md:74]: So, layer of a kind activation shares a memory block and they will be in-placed as well. Am i right?

**Review by zhoonit — APPROVED** [review]
LGTM with some minor comments

File: docs/memory-management.md (hunk 4 above)
Comment (kparichay) [reply docs/memory-management.md:74]: All activation layers share the same memory block. Only some layers are in-placed.

File: docs/memory-management.md (hunk 2 above)
Comment (kparichay) [reply docs/memory-management.md:36]: added

File: docs/memory-management.md (hunk 3 above)
Comment (kparichay) [reply docs/memory-management.md:52]: done.

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: docs/memory-management.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+Memory management for a model consists of multiple parts including:
+
+- **Weights** which make the model
+- **Input, Output, and Label tensors** for the model
+- **Variables** required to run the model like intermediate inputs and outputs, derivatives, and gradients.
+
+The management for the above 3 kinds of memory is done depending on the mode of execution of the model, namely `inference` or `training` mode. The memory management of these tensors is performed by the `Manager` inside `NNTrainer`.
+
+The memory for the `weights` are allocated at the time of initializing the model, where the `weights` are either initialized using the provided initializer or loaded from the saved model file. The memory allocated for `weights` is freed
+upon `destruction` of the model object containing the `weights`.
```
Comment (lhs8928) [inline docs/memory-management.md:16]: Concatenate the line 15, 16 will be better

**Review by lhs8928 — APPROVED** [review]
LGTM

File: docs/memory-management.md (hunk 5 above)
Comment (kparichay) [reply docs/memory-management.md:16]: I will do a new PR since this is merged.

**Pull request merged**